gen2-rgbd-projection/main.py

```diff
@@ -0,0 +1,330 @@
+    cam_coords_homogeneous = np.vstack(
+        (cam_coords, np.ones((1, cam_coords.shape[1]))))
+    depth_points_homogeneous = np.dot(
+        depth_to_rgb_extrinsic, cam_coords_homogeneous)
+
+    # project the 3d depth points onto the rgb image plane
+    rgb_frame_ref_cloud = depth_points_homogeneous[:3, :]
+    rgb_frame_ref_cloud_normalized = rgb_frame_ref_cloud / \
+        rgb_frame_ref_cloud[2, :]
+    rgb_image_pts = np.matmul(rgb_intrinsic, rgb_frame_ref_cloud_normalized)
+    rgb_image_pts = rgb_image_pts.astype(np.int16)
+    u_v_z = np.vstack((rgb_image_pts, rgb_frame_ref_cloud[2, :]))
+    lft = np.logical_and(0 <= u_v_z[0], u_v_z[0] < rgb_width)
+    rgt = np.logical_and(0 <= u_v_z[1], u_v_z[1] < rgb_height)
+    idx_bool = np.logical_and(lft, rgt)
+    u_v_z_sampled = u_v_z[:, np.where(idx_bool)]
+    y_idx = u_v_z_sampled[1].astype(int)
+    x_idx = u_v_z_sampled[0].astype(int)
+
+    # place the valid aligned points into a new depth image
+    aligned_depth_image = np.full((rgb_height, rgb_width), 0, dtype=np.uint16)
+    aligned_depth_image[y_idx, x_idx] = u_v_z_sampled[3]*depth_scale
+    return aligned_depth_image
+
+
+def quantizeDisparityFrame(frame, max_disparity):
+    """"
+    Further quantize the depth image for nice visualization
+    """
+    color_levels = 16
+    disp = (frame.astype(float).copy() *
+            (color_levels / max_disparity)).astype(np.uint8)
+    disp = (255 * disp.astype(float) / color_levels).astype(np.uint8)
+    disp = cv2.applyColorMap(disp, cv2.COLORMAP_JET)
+    return disp
+
+
+if __name__ == "__main__":
+    print("Initialize pipeline")
+    print("Align Depth: {}".format(align_depth))
+    print("Visualize Pointcloud: {}".format(not no_pcl))
+    pipeline, streams, max_disparity = create_rgbd_pipeline()
+
+    # Connect to device and start pipeline
+    print("Opening device")
+    with dai.Device(pipeline) as device:
+        # get the camera calibration info
+        calibData = device.readCalibration()
+        left_intrinsic = np.array(calibData.getCameraIntrinsics(
+            dai.CameraBoardSocket.LEFT, res["width"], res["height"]))
+        right_intrinsic = np.array(calibData.getCameraIntrinsics(
+            dai.CameraBoardSocket.RIGHT, res["width"], res["height"]))
+        right_distortion = np.array(
+            calibData.getDistortionCoefficients(dai.CameraBoardSocket.RIGHT))
+        _, rgb_default_width, rgb_default_height = calibData.getDefaultIntrinsics(
+            dai.CameraBoardSocket.RGB)
+        rgb_orig_intrinsic = np.array(
+            calibData.getCameraIntrinsics(dai.CameraBoardSocket.RGB))
+        # update camera intrinsics based on new image size
+        rgb_intrinsic = rgb_orig_intrinsic.copy()
+        width_scaling = res["width"] / rgb_default_width
+        height_scaling = res["height"] / rgb_default_height
+        rgb_intrinsic[0, :] *= width_scaling
+        rgb_intrinsic[1, :] *= height_scaling
```

Comment:
focal length will also be different when you scale it. use `calibData.getCameraIntrinsics(dai.CameraBoardSocket.RGBm resize_width, resized_height)`

Reply:
This scales the first and second row of the rgb_intrinsic matrix (which would include the x/y focal lengths). I believe that getCameraIntrinsics does not behave properly if resize_width / original_rgb_width != resize_height / original_rgb_height (i.e. if the resize causes a change in image aspect ratio). That is why I am scaling the focal length and optical center manually here.

Reply:
:point_up: This theory is based on my testing (using getCameraIntrinsics as you suggested does not work properly) and from reading the source code, which does not seem to handle non-fixed aspect ratio resizing: https://github.com/luxonis/depthai-core/blob/8ab5bf954e3b65f6ec0933e372777aac10b81900/src/device/CalibrationHandler.cpp#L197 But please let me know if I am off base here.

Reply:
I think you are right. I just scales based on one of them. If possible feel free to create a PR there. Or else I will add it to my tasks. Sorry for the delay.